File: Minimum_Size_Subarray_Sum_v1.kt
// Given an array of n positive integers and a positive integer s, find the minimal
//  length of a contiguous subarray of which the sum ≥ s. If there isn't one, return 0 instead.
class Solution {
    fun minSubArrayLen(s: Int, nums: IntArray): Int {
        if (nums.isEmpty()) return 0
        val preSum = IntArray(nums.size)
        preSum[0] = nums[0]
        for (i in 1..preSum.lastIndex) {
            preSum[i] = preSum[i - 1] + nums[i]
        }
        var res = Int.MAX_VALUE
        for (i in nums.indices) {
            for (j in i..nums.lastIndex) {
                if (preSum[j] - preSum[i] + nums[i] >= s) {
                    res = Math.min(j-i + 1, res)
                    break
                }
            }
        }

        return if (res == Int.MAX_VALUE) 0 else res
    }
}

fun main() {
    val solution = Solution()
    val testset = arrayOf(
        intArrayOf(2, 3, 1, 2, 4, 3),
        intArrayOf(2, 3, 1, 2, 4, 3),
        intArrayOf(2, 3, 1, 2, 4, 3),
        intArrayOf(1, 2, 3, 4, 5)
    )
    val ss = intArrayOf(7, 8, 11, 11)
    for (i in testset.indices) {
        println(solution.minSubArrayLen(ss[i], testset[i]))
    }
}

File: Minimum_Size_Subarray_Sum_v1.cpp
// Given an array of positive integers nums and a positive integer target,
// return the minimal length of a subarray whose sum is greater than or equal to target.
// If there is no such subarray, return 0 instead.

#include <iostream>
#include <vector>
using namespace std;
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int left = 0, right = 0;
        int sum = 0;
        int minLen = nums.size() + 1;
        while (right < nums.size()) {
            if (sum < target) {
                sum += nums[right];
                right++;
            }
            while (sum >= target && left < right) {
                minLen = min(right - left, minLen);
                sum -= nums[left];
                left++;
            }
            // cout << left << " " << right << " " << sum << endl;
        }
        if (minLen > nums.size()) {
            return 0;
        }
        return minLen;
    }
};

int main() {
    struct TestCase {
        vector<int> nums;
        int target;
    };

    vector<TestCase> testCases = {{{2, 3, 1, 2, 4, 3}, 7}, {{1, 1, 1, 1, 1, 1, 1, 1}, 11}};
    Solution sol;
    for (auto& testCase : testCases) {
        cout << sol.minSubArrayLen(testCase.target, testCase.nums) << endl;
    }
    return 0;
}